synapse/federation/sender/__init__.py

```diff
@@ -328,42 +340,57 @@ async def handle_event(event: EventBase) -> None:
                         )
                     }
 
+                    destinations.discard(self.server_name)
+
                     if send_on_behalf_of is not None:
                         # If we are sending the event on behalf of another server
                         # then it already has the event and there is no reason to
                         # send the event to it.
                         destinations.discard(send_on_behalf_of)
 
-                    logger.debug("Sending %s to %r", event, destinations)
-
                     if destinations:
-                        await self._send_pdu(event, destinations)
-
                         now = self.clock.time_msec()
                         ts = await self.store.get_received_ts(event.event_id)
 
                         synapse.metrics.event_processing_lag_by_event.labels(
                             "federation_sender"
                         ).observe((now - ts) / 1000)
 
-                async def handle_room_events(events: Iterable[EventBase]) -> None:
+                        return destinations
+                    return None
+
+                async def get_federatable_events_and_destinations(
+                    room_id: str, events: Iterable[EventBase]
+                ) -> Tuple[str, List[Tuple[EventBase, Collection[str]]]]:
                     with Measure(self.clock, "handle_room_events"):
-                        for event in events:
-                            await handle_event(event)
+                        # Get destinations for events, skip if get_destinations_for_event returns None
+                        return room_id, [
+                            (event, dests)
+                            for (event, dests) in [
+                                (event, await get_destinations_for_event(event))
+                                for event in events
+                            ]
+                            if dests is not None
+                        ]
 
                 events_by_room = {}  # type: Dict[str, List[EventBase]]
                 for event in events:
                     events_by_room.setdefault(event.room_id, []).append(event)
 
-                await make_deferred_yieldable(
+                per_room_events_and_dests = await make_deferred_yieldable(
                     defer.gatherResults(
                         [
-                            run_in_background(handle_room_events, evs)
-                            for evs in events_by_room.values()
+                            run_in_background(
+                                get_federatable_events_and_destinations, rid, evs
+                            )
+                            for rid, evs in events_by_room.items()
                         ],
                         consumeErrors=True,
                     )
-                )
+                )  # type: List[Tuple[str, List[Tuple[EventBase, Collection[str]]]]]
```

Comment:
I think that'll have a small performance hit, but I think I can do that. :+1:

Reply:
I believe that parallelising per event will cause a bigger performance hit, in this regard, not because it will have to process 100 tasks concurrently suddenly (which i believe python is able to elegantly process), but because all of those events will be requesting `prev_events` and the like, which might be cached (or cachable), and many events concurrently "fighting" for this might mean more fetches than necessary, the (rough) linearisation per room, then, is intended to keep coroutines from righting over cache-misses, and to instead let every `prev_event` request lead into the next ones for subsequent events.

Just a note, i'll resolve this thread, but i just wanted to jot down my thoughts.